### PR TITLE
[Feature] Add bfloat16 (bf16) support

### PR DESCRIPTION
**Pull request by yaox12**
## Description

Add bfloat16 support for CUDA >= 11.0.

1. Add bf16 specializations for supported functions.
2. Change the float type dispatcher from bits to real data types.
3. Make PyTorch custom autograd modules work with both bf16 and fp16.
4. Enable bf16 tests for `segment_mm` and `gather_mm`.

**Note**: `-DUSE_FP16` is removed. FP16 is enabled by default while BF16 is enabled for CUDA >= 11.0, which follows the PyTorch convention.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Examples

1. Message-passing with bf16 is supported for CUDA >= 11.0 (even on older GPUs with SM < 80). For example, the following code works well on V100.
```python
>>> import torch
>>> import dgl
>>> import dgl.function as fn
>>> dev = torch.device('cuda')
>>> g = dgl.rand_graph(30, 100).to(dev)  # Create a graph on GPU w/ 30 nodes and 100 edges.
>>> g.ndata['h'] = torch.rand(30, 16).to(dev, torch.bfloat16)  # Create bf16 node features.
>>> g.edata['w'] = torch.rand(100, 1).to(dev, torch.bfloat16)  # Create bf16 edge features.
>>> # Use DGL's built-in functions for message passing on bf16 features.
>>> g.update_all(fn.u_mul_e('h', 'w', 'm'), fn.sum('m', 'x'))
>>> g.ndata['x'].dtype
torch.bfloat16
>>> g.apply_edges(fn.u_dot_v('h', 'x', 'hx'))
>>> g.edata['hx'].dtype
torch.bfloat16
```

2. AMP only works on GPUs with SM >= 80 and CUDA >= 11.0.

This also follows the PyTorch convention. See https://github.com/dmlc/dgl/issues/4333#issuecomment-1253277674.

### CODE REVIEW

File: src/array/cuda/bf16.cuh [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+__device__ __forceinline__ __nv_bfloat16  operator--(__nv_bfloat16& h, int) {  // NOLINT
+  __nv_bfloat16 ret = h; h = __nv_bfloat16(float(h) - 1.0f); return ret;  // NOLINT
+}
+
+__device__ __forceinline__ __nv_bfloat16 operator+(const __nv_bfloat16& h) { return h; }
+__device__ __forceinline__ __nv_bfloat16 operator-(const __nv_bfloat16& h) {
+  return __nv_bfloat16(-float(h));  // NOLINT
+}
+
+__device__ __forceinline__ bool operator==(const __nv_bfloat16& lh, const __nv_bfloat16& rh) {
```
Comment (frozenbugs) [inline src/array/cuda/bf16.cuh:98]: I think we are following Google C++ code style. 
Which advocate 80 characters line.
https://google.github.io/styleguide/cppguide.html#Line_Length

**Comment by frozenbugs** [comment]
Thanks for the great work, even with several comment, this PR still hold high standard code quality :)
Another comment I would suggest you to do in the future is to avoid such a big PR, if possible, separate the refactoring and adding new lines of code such that it is easier to be reviewed.

**Review by isratnisa — CHANGES_REQUESTED** [review]
This is a huge effort. Thanks! Have you done any regression test? How is the performance compare to fp16? I have left a few comments about disabling CPU support for half precision. Otherwise, LGTM.

**Review by ndickson-nvidia — CHANGES_REQUESTED** [review]

File: include/dgl/runtime/ndarray.h [hunk 2]
```diff
@@ -18,8 +18,15 @@
 #include "shared_mem.h"
 
 #ifdef DGL_USE_CUDA
+#include <cuda_runtime.h>
+
+#define BF16_ENABLED defined(CUDART_VERSION) && CUDART_VERSION >= 11000
```
[the remark this answers is not in the document]
Comment (yaox12) [reply include/dgl/runtime/ndarray.h]: Changed accordingly.

File: src/array/cuda/utils.h [hunk 3]
```diff
@@ -170,15 +140,8 @@ __global__ void _LinearSearchKernel(
...
-      // GCC treats int64_t as a distinct type from long long, so
-      // without the explcit cast to long long, it errors out saying
-      // that the implicit cast results in an ambiguous choice of
-      // constructor for __half.
-      // The using statement is to avoid a linter error about using
-      // long or long long.
-      using LongLong = long long; // NOLINT
-      out[tx] = weights ? weights[v] : DType(LongLong(v));
+      // The casts here are to be able to handle DType being __half or __nv_bfloat16.
+      out[tx] = weights ? weights[v] : static_cast<DType>(static_cast<double>(v));
```
[the remark this answers is not in the document]
Comment (yaox12) [reply src/array/cuda/utils.h]: Added a specialization for bfloat16.

File: src/array/cpu/gather_mm.cc [hunk 4]
```diff
@@ -62,74 +48,52 @@ void GatherMMScatter(const NDArray A,
     LOG(FATAL) << "Unsupported CPU kernel for GatherMM.";
 }
 
-template void GatherMM<kDGLCPU, int32_t, 16>(
```
[the remark this answers is not in the document]
Comment (yaox12) [reply src/array/cpu/gather_mm.cc:65]: Yes. In the current master branch, we either fall back to float32 or throw an error for half precision on CPU. I think it's OK to remove it.

**Comment by yaox12** [comment]
> Have you done any regression test?

Not yet. It seems I don't have the authentication to trigger regression tests, and I failed to run it locally.
 
> How is the performance compare to fp16?

For the AMP example in https://docs.dgl.ai/en/0.9.x/guide/mixed_precision.html, bf16 has a similar performance as fp16, both slightly faster than fp32. 
An advantage of bf16 is that it supports pure-bf16 training without AMP. Thus the example in https://github.com/dmlc/dgl/pull/4262 could converge with bf16 but not fp16.

File: src/array/cuda/atomic.cuh [hunk 5]
```diff
@@ -275,7 +301,24 @@ __device__ __forceinline__ half AtomicAdd<half>(half* addr, half val) {
 #endif  // __CUDA_ARCH__ >= 700
 }
 #endif  // defined(CUDART_VERSION) && CUDART_VERSION >= 10000
-#endif  // USE_FP16
+
+#if BF16_ENABLED
+template <>
+__device__ __forceinline__ __nv_bfloat16 AtomicAdd<__nv_bfloat16>(
+    __nv_bfloat16* addr, __nv_bfloat16 val) {
+// half make sure we have half support
```
[the remark this answers is not in the document]
Comment (yaox12) [reply src/array/cuda/atomic.cuh]: Update. This is a copy-paste typo. Should be "make sure we have bfloat16 support".

File: src/array/cuda/atomic.cuh [hunk 6]
```diff
@@ -54,7 +54,32 @@ template <> struct Cast<half> {
...
 };
+
+#if BF16_ENABLED
+template <> struct Cast<__nv_bfloat16> {
+  typedef Code<sizeof(__nv_bfloat16)>::Type Type;
+  static __device__ __forceinline__ Type Encode(__nv_bfloat16 val) {
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 800
+    return __bfloat16_as_ushort(val);
+#else
+  printf("Atomic operations are not supported for bfloat16 (BF16) "
```
[the remark this answers is not in the document]
Comment (yaox12) [reply src/array/cuda/atomic.cuh]: Update accordingly.

[the remark this answers is not in the document]
Comment (yaox12) [reply src/array/cuda/atomic.cuh]: This is a device(GPU)-side error message, which doesn't support the `LOG` macro. So we just use `printf` as a workaround.

[the remark this answers is not in the document]
Comment (frozenbugs) [reply src/array/cuda/atomic.cuh]: Looks like an AI to take in the future, not sure whether you are happy to own it or not, we can discuss it in the meeting.

[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/cuda/atomic.cuh]: Could we move the check outside the device code though?

**Review by isratnisa — APPROVED** [review]
